File: Structural/Composite/Composite.py
from __future__ import annotations
from abc import ABC, abstractmethod
from typing import List, Optional

# 1. 统一组件接口
class Graphic(ABC):
    @abstractmethod
    def draw(self) -> None: ...

    def add(self, g: "Graphic") -> None:
        raise NotImplementedError
    def remove(self, g: "Graphic") -> None:
        raise NotImplementedError
    def get_child(self, idx: int) -> Optional["Graphic"]:
        raise NotImplementedError

# 2. 叶子节点
class Line(Graphic):
    def draw(self) -> None:
        print("Draw Line")

class Text(Graphic):
    def draw(self) -> None:
        print("Draw Text")

# 3. 组合节点
class Picture(Graphic):
    def __init__(self) -> None:
        self._children: List[Graphic] = []

    def draw(self) -> None:
        print("--- Picture Start ---")
        for child in self._children:
            child.draw()
        print("--- Picture End ---")

    def add(self, g: Graphic) -> None:
        self._children.append(g)

    def remove(self, g: Graphic) -> None:
        self._children.remove(g)

    def get_child(self, idx: int) -> Optional[Graphic]:
        return self._children[idx] if 0 <= idx < len(self._children) else None

# 4. Demo
if __name__ == "__main__":
    root = Picture()
    root.add(Line())
    root.add(Text())

    sub = Picture()
    sub.add(Line())
    sub.add(Text())
    root.add(sub)

    root.draw()   # 递归绘制整棵树

File: Structural/Composite/Composite.java
import java.util.*;

// 1. 统一组件接口
public interface Graphic {
    void draw();
    default void add(Graphic g) { throw new UnsupportedOperationException(); }
    default void remove(Graphic g) { throw new UnsupportedOperationException(); }
    default Graphic getChild(int i) { throw new UnsupportedOperationException(); }
}

// 2. 叶子节点：Line
class Line implements Graphic {
    public void draw() { System.out.println("Draw Line"); }
}

// 3. 叶子节点：Text
class Text implements Graphic {
    public void draw() { System.out.println("Draw Text"); }
}

// 4. 组合节点：Picture
class Picture implements Graphic {
    private final List<Graphic> children = new ArrayList<>();

    public void draw() {
        System.out.println("--- Picture Start ---");
        children.forEach(Graphic::draw);
        System.out.println("--- Picture End ---");
    }

    public void add(Graphic g) { children.add(g); }
    public void remove(Graphic g) { children.remove(g); }
    public Graphic getChild(int i) { return children.get(i); }
}

// 5. Demo
class CompositeDemo {
    public static void main(String[] args) {
        Graphic root = new Picture();
        root.add(new Line());
        root.add(new Text());

        Graphic sub = new Picture();
        sub.add(new Line());
        sub.add(new Text());
        root.add(sub);

        root.draw();   // 一次性递归绘制整棵树
                    //  调用5次draw() root.draw() -> root.children[0].draw() , root.children[1].draw() , root.children[2].draw()->root.children[2][0].draw , root.children[2][1].draw()
                    //                                  Line() , Text() , Picture()  ;而Picture() : Line()  Text()  
     }
}
